Initial implementation of filtered full composite image

trylek opened the pull request:
Based on startup perf measurement experiments I have identified an initial subset of framework + ASP.NET assemblies that are the most critical for ASP.NET app startup; by filtering composite native library build to only R2R precompile code from these, not from all component assemblies, we believe to be able to reach startup perf comparable with default R2R publishing but with publishing size smaller by about 35% (on x64 the default R2R publishing size of the ASP.NET benchmark app is about 111 MB, with this change it is about 70 MB).

Our current expectation is that after receiving customer feedback we'll decide whether to proceed in this direction and potentially introduce a more robust and repeatable way of periodically updating the filtered assembly list used for composite size optimization.

Thanks

Tomas

trylek committed:
Initial implementation of filtered full composite image

Based on startup perf measurement experiments I have identified
an initial subset of framework + ASP.NET assemblies that are
the most critical for ASP.NET app startup; by filtering composite
native library build to only R2R precompile code from these,
not from all component assemblies, we believe to be able to reach
startup perf comparable with default R2R publishing but with
publishing size smaller by about 35% (on x64 the default R2R
publishing size of the ASP.NET benchmark app is about 111 MB,
with this change it is about 70 MB).

Our current expectation is that after receiving customer feedback
we'll decide whether to proceed in this direction and potentially
introduce a more robust and repeatable way of periodically
updating the filtered assembly list used for composite size
optimization.

Thanks

Tomas

Finetune the filtering assembly list

trylek commented:
/cc @dotnet/crossgen-contrib

wtgodbe reviewed (approved):
The changes seem reasonable to me, is there any noticeable difference in size for the Shared Framework package?

ivdiazsa commented on src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj:
```diff
@@ -483,22 +483,39 @@ This package is an internal implementation of the .NET Core SDK and is not meant
+      <PartialCompositeAssemblyPaths Include="$(ManagedAssetsFullPath)\%(PartialCompositeAssemblyNames.Identity).dll" />
+      <PartialCompositeAssemblyPaths Include="$(TargetDir)\%(PartialCompositeAssemblyNames.Identity).dll" />
+      <PartialCompositeAssemblyPaths Remove="%(PartialCompositeAssemblyPaths.Identity)" Condition="!Exists('%(Identity)')" />
+      <UnrootedAssemblyPaths Include="@(BaseAssemblies)" />
```
I'm confused. What is this `UnrootedAssemblyPaths`? It seems to be a list of the assemblies that will not be part of the composite. Why do we need it?

trylek replied:
The logic I'm using here is slightly different than the measurements you were doing last year. In your implementation you were creating the composite image from just a subset of the framework / ASP.NET assemblies and the remaining assemblies were left in their original shipping (R2R-compiled) form. This approach uses the "full filtered composite" where we make a composite comprising all framework and ASP.NET assemblies but not all of them generate R2R code into the image; in particular, for "unrooted inputs", we don't mark all their methods as roots - they get compiled if they're reachable from another assembly via dependency analysis, they're just not compiled automatically. This filtering is what lets us reduce the total publishing size while keeping startup time on par with default R2R publishing - for some assemblies that are rarely used we drop the R2R code altogether in this scenario.

ivdiazsa replied:
Ok thanks for the explanation Tomas. Then I think this is ready to be merged.

trylek commented:
@wtgodbe - Unfortunately I haven't yet figured out whether there's a way to inspect the nupkg generated by the aspnetcore CI build, I haven't found it in the artifacts so the numbers I'm going to present are based on my local builds with the caveat that manual steps are required so something may be slightly off. In addition to that, this form of compound composite image comprises both ASP.NET and the framework so there's no easy way to refer to "just the shared framework package" without ASP.NET. From what I'm seeing when building the ASP.NET Benchmarks app as self-contained targeting Linux x64:

**dotnet publish -c Release --sc -f net8.0 -r linux-x64 /p:PublishReadyToRun=true** - "publish" folder size = 136 MB (including native CoreCLR runtime and other support stuff).

**dotnet publish -c Release --sc -f net8.0 -r linux-x64 /p:PublishREadyToRun=true /p:PublishReadyToRunComposite=true" - "publish" folder size = 157 MB.

**dotnet publish -c Release --sc f net8.0 -r linux-x64 /p:PublishReadyToRun=true followed by manually replacing the framework and ASP.NET with this new form of composite** - "publish" folder size = 101 MB.

Thanks

Tomas

trylek commented on src/Framework/App.Runtime/src/PartialCompositeAssemblyList.txt:
```diff
@@ -0,0 +1,53 @@
+Microsoft.AspNetCore.Server.Kestrel.Transport.Quic
+System.Net.Quic
+System.Diagnostics.Tracing
+Microsoft.Win32.Registry
```
That's true, I have mostly used Windows for updating the training assembly lists with the assumption that the diff shouldn't be too big, I can retry using Linux but I don't expect the difference to be huge.

mangod9 replied:
yeah not part of this change but we should try to remove this if possible. Even on windows wouldnt think the runtime should be looking into the registry too much.

trylek replied:
No problem, I can generate the list on Linux and retest the perf, that's mostly trivial by now for me.

trylek replied:
I have updated the list by rerunning the training tests on Linux based on your suggestion in the 2nd commit.

mangod9 replied:
ok nice, assume the same size gains are noticeable with this new list?

trylek replied:
Yes I believe so. In fact when I was updating the assembly filtering list based on the Linux results, I was slightly more aggressive in pruning the list - while I'm still seeing startup perf on par with or slightly better than for default R2R publishing, the difference in size of the managed assemblies (composite + components - the individual R2R files in default publishing) is about -40 MB (108 MB in default R2R publishing case, 68 MB in the filtered full composite case). I believe some headroom may be useful if we or customers identify additional assemblies that can be hot at startup in certain mainstream scenarios.

wtgodbe commented:
> Unfortunately I haven't yet figured out whether there's a way to inspect the nupkg generated by the aspnetcore CI build

We don't produce them by default, but your local numbers look very promising! If you want to download an "official" package you could push this branch to the [internal repo](https://dev.azure.com/dnceng/internal/_git/dotnet-aspnetcore) then run an [official build](https://dev.azure.com/dnceng/internal/_build?definitionId=21) against that branch, but I'm happy to sign off without that.

trylek committed:
Adjust the assembly filtering list based on Manish's PR feedback

I have updated the assembly list by rerunning the training tests
on Linux; based on the perf results I tried to be even slightly
more aggressive in trimming of the composite image.

Thanks

Tomas

trylek commented:
Thanks everyone for your feedback. I have triggered a new round of tests mostly to double-check whether I'm not hitting a conflict with Ivan's musl fix, that is apparently not the case, I plan to merge this in once the tests finish unless anyone objects.
